From david-b@pacbell.net  Thu Jul  3 16:17:33 2008
From: David Brownell <david-b@pacbell.net>
Date: Tue, 1 Jul 2008 13:18:20 -0700
Subject: usb_gadget: composite cdc gadget fault handling
To: Greg KH <greg@kroah.com>
Cc: linux-usb@vger.kernel.org
Message-ID: <200807011318.20220.david-b@pacbell.net>
Content-Disposition: inline


These two fixes ensure the new "CDC Composite Device" gadget
fails cleanly when it's loaded on hardware that can't support
this particular gadget driver.

Signed-off-by: David Brownell <dbrownell@users.sourceforge.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/cdc2.c     |    2 +-
 drivers/usb/gadget/u_serial.c |    3 +++
 2 files changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/usb/gadget/cdc2.c
+++ b/drivers/usb/gadget/cdc2.c
@@ -124,7 +124,7 @@ static int __init cdc_do_config(struct u
 		return status;
 
 	status = acm_bind_config(c, 0);
-	if (status == 0)
+	if (status < 0)
 		return status;
 
 	return 0;
--- a/drivers/usb/gadget/u_serial.c
+++ b/drivers/usb/gadget/u_serial.c
@@ -1090,6 +1090,9 @@ void gserial_cleanup(void)
 	unsigned	i;
 	struct gs_port	*port;
 
+	if (!gs_tty_driver)
+		return;
+
 	/* start sysfs and /dev/ttyGS* node removal */
 	for (i = 0; i < n_ports; i++)
 		tty_unregister_device(gs_tty_driver, i);
